### ISSUE TITLE
assertion failed: ddp->dd_ref > 1 in dls_mgmt.c:1276

### ISSUE THREAD

**Issue by wiedi** [TritonDataCenter/illumos-joyent#155]
We hit the following panic:

```
> $C
ffffff000fcd4a50 vpanic()
ffffff000fcd4a80 0xfffffffffba7d538()
ffffff000fcd4ab0 dls_devnet_rele+0x70(ffffff03cdceb2e8)
ffffff000fcd4af0 dld_str_detach+0x79(ffffff03ea58fd20)
ffffff000fcd4b30 dld_str_close+0x6a(ffffff03e8bc2530)
ffffff000fcd4b60 dld_close+0x20(ffffff03e8bc2530)
ffffff000fcd4be0 qdetach+0x9b(ffffff03e8bc2530, 1, 3, ffffff03eac46260, 0)
ffffff000fcd4c80 strclose+0x2c1(ffffff041b90cd00, 3, ffffff03eac46260)
ffffff000fcd4cd0 device_close+0xb9(ffffff04163f7880, 3, ffffff03eac46260)
ffffff000fcd4d60 spec_close+0x17b(ffffff04163f7880, 3, 1, 0, ffffff03eac46260, 0)
ffffff000fcd4de0 fop_close+0x61(ffffff04163f7880, 3, 1, 0, ffffff03eac46260, 0)
ffffff000fcd4e20 closef+0x5e(ffffff03cba35e90)
ffffff000fcd4e90 closeandsetf+0x398(3, 0)
ffffff000fcd4eb0 close+0x13(3)
ffffff000fcd4f10 sys_syscall32+0x12f()
> ::status
debugging crash dump vmcore.0 (64-bit) from cube-b
operating system: 5.11 skylime_20171203T161229Z (i86pc)
image uuid: (not set)
panic message: assertion failed: ddp->dd_ref > 1, file: ../../common/io/dls/dls_mgmt.c, line: 1276
dump content: kernel pages only

```

What triggered it was something like this:

- window 1: `snoop -d net0 -z <uuid>`
- window 2: `vmadm stop -F <uuid>`
- window 1: CTRL+C

I can share the vmdump if that helps. This panic was on a custom SmartOS build but I'm relatively sure we hit the same/similar panic a few days ago on a joyent_20170608T172228Z box. Sadly I could not get a dump of that.

I think there are still some changes to DLS in the joyent tree that are not yet upstream, so I'm not sure if this applies to upstream illumos-gate or only to illumos-joyent.

**Comment by danmcd** [comment]
The assert in question was introduced here:

https://github.com/joyent/illumos-joyent/commit/87b3eba89974d328822c6ffda4259b7f3d4edc54#diff-d892ff431ea6494d85f7915fcf3aea88

I have my doubts the commit itself CAUSED this bug.  I suspect the VERIFY() here exposes an older DLS bug.  I know there's active work cleaning up messes here.  Thanks for this report, and if you can share the vmdump, we'd very much appreciate it.

**Comment by rzezeski** [comment]
@wiedi Thanks for the report.

This is a known issue: [OS-6340](https://smartos.org/bugview/OS-6340). I know the fix but I haven't implemented it yet because it may have ripple effects on the zone shutdown code and I'm busy with some other stuff at the moment.

This should only affect illumos-joyent.

**Comment by wiedi** [comment]
Ah, interesting writeup. Glad you've already figured it out.

**Comment by wiedi** [comment]
We hit this today again after issuing a `vmadm delete` on joyent_20230223T001442Z.

```
reboot after panic: assertion failed: ddp->dd_ref > 1, file: ../../common/io/dls/dls_mgmt.c, line: 1282
```

Sadly no usable dump.

After having a look around I assume the fix is 85dff7a05711e1238299281f8a94d2d40834c775.
So we should update to a more recent version that includes it and this issue can be closed, right? :)
